src/ex_color/model.py

```diff
@@ -44,45 +44,72 @@ def forward(self, x: Tensor) -> Tensor:
         return self.decoder(latents)
 
 
-class LatentCaptureHook:
-    """Hook class to capture latent representations externally."""
+class ActivationCaptureHook:
+    """Hook class to capture latent representations."""
 
     def __init__(self):
-        self.current_latents: Tensor | None = None
+        self.activations: Tensor | None = None
 
     def __call__(self, _module, _input, output):
-        """Hook function to capture latent representations."""
-        self.current_latents = output
+        self.activations = output
 
 
 class Objective(Protocol):
     def __call__(self, y_pred: Tensor, y_true: Tensor) -> Tensor: ...
 
 
-class ColorMLPTrainingModule(L.LightningModule):
-    """Lightning module that handles training logic."""
+class TrainingModule(L.LightningModule):
+    """Lightning module that handles training logic for any model architecture."""
 
     def __init__(
         self,
+        model: nn.Module,
         dopesheet: Dopesheet,
         objective: Objective,
         regularizers: list[RegularizerConfig],
     ):
         super().__init__()
-        self.save_hyperparameters()
+
+        # Save params for rehydration. Modules are saved automatically.
+        ignore = ['model']
+        if isinstance(objective, nn.Module):
+            ignore.append('objective')
+        self.save_hyperparameters(ignore=ignore)
 
         # Store training configuration
         self.objective = objective
         self.regularizers = regularizers
         self.timeline = Timeline(dopesheet)
 
-        # Create the pure neural network model (without training logic)
-        self.model = ColorMLP()
+        # Store the pure neural network model (passed as parameter)
+        self.model = model
+
+        # Set up latent capture hooks for all unique layer names
+        self.latent_hooks: dict[str, ActivationCaptureHook] = {}
+        self._setup_latent_hooks()
```

Comment:
Do you think this should be done in a lifecycle hook, like `on_train_start`/`on_fit_start`? And then, should we call `handle.remove()` in `on_train_end`/`on_fit_end`? I'm not super familiar with Lightning so your thoughts on this would be appreciated.

Docs:

> `LightningModule.on_train_start()`
>
> Called at the beginning of training after sanity check.

> `LightningModule.on_train_end()`
>
> Called at the end of training before logger experiment is closed.

> `LightningModule.on_fit_start()`
>
> Called at the very beginning of fit. If on DDP it is called on every process

> `LightningModule.on_fit_end()`
>
> Called at the very end of fit. If on DDP it is called on every process

I guess it should probably be called on every process?